test/unit/host_counter_test.rb

```diff
@@ -0,0 +1,64 @@
+require "test_helper"
+
+class HostCounterTest < ActiveSupport::TestCase
+  def setup
+    User.current = users :admin
+  end
+
+  test 'it should get number of hosts associated to model' do
+    m = Model.create(:name => "model-1")
+    Host.create(:model => m)
+    count = HostCounter.new(:model)
+    assert_equal 1, count[m]
+  end
+
+  test 'it should get number of hosts associated to architecture' do
+    os = Architecture.create(:name => "arch-1")
+    Host.create(:architecture => os)
+    Host.create(:architecture => os)
+    count = HostCounter.new(:architecture)
+    assert_equal 2, count[os]
+  end
+
+  test 'it should count only hosts if user has view_hosts permissions' do
+    model = Model.first
+    assert Host.create(:model => model)
+
+    # admin has view_hosts permissions
+    count = HostCounter.new(:model)
+    assert_equal 1, count[model]
+
+    # user "one" does not have view hosts permissions
+    User.current = users :one
```

Comment:
you can use `as_user(:one) do...` or `as_admin do...` instead of manually changing the current user